File: libs/ui_widgets/src/core/tree_sync.cpp
#include "pwb/ui_widgets/core/tree_sync.hpp"

#include <cstdlib>

namespace pwb::ui_widgets::core {

bool py_truthy(const nlohmann::json& value) {
    switch (value.type()) {
        case nlohmann::json::value_t::null:
            return false;
        case nlohmann::json::value_t::boolean:
            return value.get<bool>();
        case nlohmann::json::value_t::number_integer:
        case nlohmann::json::value_t::number_unsigned:
            return value.get<std::int64_t>() != 0;
        case nlohmann::json::value_t::number_float:
            return value.get<double>() != 0.0;
        case nlohmann::json::value_t::string:
            return !value.get_ref<const std::string&>().empty();
        case nlohmann::json::value_t::array:
        case nlohmann::json::value_t::object:
        case nlohmann::json::value_t::binary:
        case nlohmann::json::value_t::discarded:
            return !value.empty();
    }
    return false;
}

std::string py_str(const nlohmann::json& value) {
    switch (value.type()) {
        case nlohmann::json::value_t::null:
            return "None";
        case nlohmann::json::value_t::boolean:
            return value.get<bool>() ? "True" : "False";
        case nlohmann::json::value_t::string:
            return value.get<std::string>();
        case nlohmann::json::value_t::number_integer:
        case nlohmann::json::value_t::number_unsigned:
        case nlohmann::json::value_t::number_float:
            // repr-shortest == Python str() for ints/floats (nlohmann dump).
            return value.dump();
        case nlohmann::json::value_t::array:
        case nlohmann::json::value_t::object:
        case nlohmann::json::value_t::binary:
        case nlohmann::json::value_t::discarded:
            // Python str(container) uses repr formatting — unreachable on the
            // wire (ids are scalars); JSON dump is the honest approximation.
            return value.dump();
    }
    return {};
}

std::int64_t py_int(const nlohmann::json& value) {
    switch (value.type()) {
        case nlohmann::json::value_t::boolean:
            return value.get<bool>() ? 1 : 0;
        case nlohmann::json::value_t::number_integer:
        case nlohmann::json::value_t::number_unsigned:
            return value.get<std::int64_t>();
        case nlohmann::json::value_t::number_float:
            return static_cast<std::int64_t>(value.get<double>());
        case nlohmann::json::value_t::string: {
            const std::string& s = value.get_ref<const std::string&>();
            char* end = nullptr;
            const long long parsed = std::strtoll(s.c_str(), &end, 10);
            if (end == s.c_str() || *end != '\0') return 0;  // int("5.5") -> ValueError
            return parsed;
        }
        default:
            return 0;
    }
}

namespace {

nlohmann::ordered_json parse_payload(const std::string& payload, bool* ok) {
    *ok = false;
    if (payload.empty()) return nlohmann::ordered_json::object();
    nlohmann::ordered_json data =
        nlohmann::ordered_json::parse(payload, nullptr, false);
    if (data.is_discarded() || !data.is_object()) {
        return nlohmann::ordered_json::object();
    }
    *ok = true;
    return data;
}

TreeChangeSet change_set_from(const nlohmann::ordered_json& data) {
    TreeChangeSet out;
    const auto visibility = data.value("visibility", nlohmann::ordered_json{});
    if (visibility.is_object()) {
        for (auto it = visibility.begin(); it != visibility.end(); ++it) {
            out.visibility[it.key()] = py_truthy(it.value());
        }
    }
    const auto order = data.value("order", nlohmann::ordered_json{});
    if (order.is_array()) {
        for (const auto& item : order) out.order.push_back(py_str(item));
    }
    const auto renames = data.value("renames", nlohmann::ordered_json{});
    if (renames.is_object()) {
        for (auto it = renames.begin(); it != renames.end(); ++it) {
            out.renames[it.key()] = py_str(it.value());
        }
    }
    return out;
}

}  // namespace

std::map<std::string, bool> TreeChangeBatch::group_visibility() const {
    std::map<std::string, bool> out;
    for (const TreeEvent& event : events) {
        if (event.type == "visibility" && event.is_group()) {
            out[event.node_id] = event.visibility_value;
        }
    }
    return out;
}

std::map<std::string, std::string> TreeChangeBatch::group_renames() const {
    std::map<std::string, std::string> out;
    for (const TreeEvent& event : events) {
        if (event.type == "rename" && event.is_group()) {
            out[event.node_id] = event.rename_value;
        }
    }
    return out;
}

TreeChangeSet parse_tree_change(const std::string& payload) {
    bool ok = false;
    const nlohmann::ordered_json data = parse_payload(payload, &ok);
    if (!ok) return TreeChangeSet{};
    return change_set_from(data);
}

TreeChangeBatch parse_tree_events(const std::string& payload) {
    bool ok = false;
    const nlohmann::ordered_json data = parse_payload(payload, &ok);
    TreeChangeBatch batch;
    if (!ok) return batch;
    batch.changes = change_set_from(data);

    // Python `str(raw.get(k) or default)` — falsy collapses to default first.
    auto str_or = [](const nlohmann::ordered_json& obj, const char* key,
                     const char* fallback) -> std::string {
        const auto value = obj.value(key, nlohmann::ordered_json{});
        return py_truthy(value) ? py_str(value) : fallback;
    };

    const auto events = data.value("events", nlohmann::ordered_json{});
    if (events.is_array()) {
        for (const auto& raw : events) {
            if (!raw.is_object()) continue;
            const std::string event_type = str_or(raw, "type", "");
            const std::string node_type = str_or(raw, "node_type", "layer");
            const std::string node_id = str_or(raw, "node_id", "");
            if ((event_type != "visibility" && event_type != "rename") ||
                node_id.empty()) {
                continue;
            }
            TreeEvent event;
            event.type = event_type;
            event.node_type = node_type;
            event.node_id = node_id;
            const nlohmann::ordered_json value =
                raw.value("value", nlohmann::ordered_json{});
            if (event_type == "visibility") {
                event.visibility_value = py_truthy(value);
            } else {
                event.rename_value = py_str(value);
            }
            batch.events.push_back(std::move(event));
        }
    }
    const auto tree = data.value("tree", nlohmann::ordered_json{});
    if (tree.is_array()) {
        for (const auto& node : tree) {
            if (node.is_object()) batch.tree.push_back(node);
        }
    }
    // `data.get("tree_revision") or 0` then int() — falsy -> 0, bad -> 0.
    const auto revision = data.value("tree_revision", nlohmann::ordered_json{});
    batch.revision = py_truthy(revision) ? py_int(revision) : 0;
    return batch;
}

}  // namespace pwb::ui_widgets::core

File: legacy/python_reference/product/paleo_workbench/ui/qgis_stack/tree_sync.py
"""树变更回调 → 文档模型写回的解析层（legacy 平铺 + V5 schema 2）。

C++ 桥 ``set_tree_change_callback`` 的 payload 契约（见
``native/qgis_render_bridge/src/map_stack_service.cpp`` 的 flushTreeChange）：

legacy 键（平铺图层语义，既有消费者保持不变）：

    {"visibility": {doc_id: bool}, "order": [doc_id...],
     "renames": {doc_id: name}}

V5 schema 2 追加：

    {"schema": 2,
     "events": [{"type": "visibility"|"rename",
                 "node_type": "layer"|"group",
                 "node_id": str, "value": bool|str}, ...],
     "tree": [{"type": "group", "id": gid, "name": n, "visible": bool,
               "children": [...]},
              {"type": "layer", "id": doc_id, ...}]}

``tree`` 只在结构变化（拖拽/建组/删组）时携带，一次性覆盖
move/group-create/group-delete——Python 侧对结构化节点数组 diff，
不做字符串拼接解析。
"""
from __future__ import annotations

import json
from dataclasses import dataclass, field


@dataclass(frozen=True)
class TreeChangeSet:
    """legacy 平铺变更批次（旧语义，保留给既有消费者）。"""

    visibility: dict[str, bool] = field(default_factory=dict)
    order: tuple[str, ...] = ()
    renames: dict[str, str] = field(default_factory=dict)

    @property
    def empty(self) -> bool:
        return not (self.visibility or self.order or self.renames)


@dataclass(frozen=True)
class TreeEvent:
    """V5 typed 树事件。"""

    type: str            # "visibility" | "rename"
    node_type: str       # "layer" | "group"
    node_id: str
    value: object        # bool（visibility）| str（rename）

    @property
    def is_group(self) -> bool:
        return self.node_type == "group"


@dataclass(frozen=True)
class TreeChangeBatch:
    """schema 2 完整批次：legacy 键 + typed events + 结构快照。"""

    changes: TreeChangeSet
    events: tuple[TreeEvent, ...] = ()
    #: 结构快照节点数组（结构变化时非空）：
    #: [{"type": "group"|"layer", "id": ..., "children": [...]}]。
    tree: tuple[dict, ...] = ()
    #: V11 树修订号（桥 0.7.0a0+ 携带；0 = 旧桥无修订号——不过期）。
    revision: int = 0

    @property
    def empty(self) -> bool:
        return self.changes.empty and not self.events and not self.tree

    @property
    def has_structure_change(self) -> bool:
        return bool(self.tree)

    def group_visibility(self) -> dict[str, bool]:
        return {
            event.node_id: bool(event.value)
            for event in self.events
            if event.type == "visibility" and event.is_group
        }

    def group_renames(self) -> dict[str, str]:
        return {
            event.node_id: str(event.value)
            for event in self.events
            if event.type == "rename" and event.is_group
        }


def parse_tree_change(payload: str) -> TreeChangeSet:
    """legacy 解析（既有行为保持不变；缺失键/坏 JSON 返回空集）。"""
    try:
        data = json.loads(payload) if payload else {}
    except (TypeError, ValueError):
        return TreeChangeSet()
    if not isinstance(data, dict):
        return TreeChangeSet()
    visibility = {
        str(key): bool(value)
        for key, value in (data.get("visibility") or {}).items()
    }
    order = tuple(str(item) for item in (data.get("order") or ()))
    renames = {
        str(key): str(value) for key, value in (data.get("renames") or {}).items()
    }
    return TreeChangeSet(visibility=visibility, order=order, renames=renames)


def parse_tree_events(payload: str) -> TreeChangeBatch:
    """schema 2 解析：legacy 键 + typed events + 结构快照。"""
    try:
        data = json.loads(payload) if payload else {}
    except (TypeError, ValueError):
        return TreeChangeBatch(changes=TreeChangeSet())
    if not isinstance(data, dict):
        return TreeChangeBatch(changes=TreeChangeSet())
    events: list[TreeEvent] = []
    for raw in data.get("events") or ():
        if not isinstance(raw, dict):
            continue
        event_type = str(raw.get("type") or "")
        node_type = str(raw.get("node_type") or "layer")
        node_id = str(raw.get("node_id") or "")
        if event_type not in ("visibility", "rename") or not node_id:
            continue
        value = raw.get("value")
        value = bool(value) if event_type == "visibility" else str(value)
        events.append(TreeEvent(event_type, node_type, node_id, value))
    tree = tuple(
        node for node in (data.get("tree") or ()) if isinstance(node, dict)
    )
    try:
        revision = int(data.get("tree_revision") or 0)
    except (TypeError, ValueError):
        revision = 0
    return TreeChangeBatch(
        changes=parse_tree_change(payload),
        events=tuple(events),
        tree=tree,
        revision=revision,
    )
